docs/architecture/adr-062-collections-state-layer.md

```diff
@@ -0,0 +1,117 @@
+# ADR 062: Collections, a simplified storage layer for cosmos-sdk modules.
+
+## Changelog
+
+* 30/11/2022: PROPOSED
+
+## Status
+
+PROPOSED - Implemented
+
+## Abstract
+
+We propose a simplified module storage layer which leverages golang generics to allow module developers to handle module
+storage in a simple and straightforward manner, whilst offering safety, extensibility and standardisation.
+
+## Context
+
+Module developers are forced into manually implementing storage functionalities in their modules, those functionalities include
+but are not limited to:
+
+- Defining key to bytes formats.
+- Defining value to bytes formats.
+- Defining secondary indexes.
+- Defining query methods to expose outside to deal with storage.
+- Defining local methods to deal with storage writing.
+- Dealing with genesis imports and exports.
+- Writing tests for all the above.
+
+
+This brings in a lot of problems:
+- It blocks developers from focusing on the most important part: writing business logic.
+- Key to bytes formats are complex and their definition is error-prone, for example:
+  - how do I format time to bytes in such a way that bytes are sorted?
+  - how do I ensure when I don't have namespace collisions when dealing with secondary indexes?
+- The lack of standardisation makes life hard for clients, and the problem is exacerbated when it comes to providing proofs for objects present in state. Clients are forced to maintain a list of object paths to gather proofs.
+
+### Current Solution: ORM
+
+The current SDK proposed solution to this problem is [ORM](https://github.com/cosmos/cosmos-sdk/blob/main/docs/architecture/adr-055-orm.md).
+Whilst ORM offers a lot of good functionality aimed at solving these specific problems, it has some downsides:
+- It requires migrations.
+- It uses the newest protobuf golang API, whilst the SDK still mainly uses gogoproto. 
+- Integrating ORM into a module would require the developer to deal with two different golang frameworks (golang protobuf + gogoproto) representing the same API objects.
+- It has a high learning curve, even for simple storage layers as it requires developers to have knowledge around protobuf options, custom cosmos-sdk storage extensions, and tooling download. Then after this they still need to learn the code-generated API.
```

Comment:
I agree there are some downsides at the moment but the downsides related to gogoproto seems more along the lines of an issue with the sdk needing to be updated rather than a downside of the orm module itself.

Also, developers should already have working knowledge of protobuf through module development and learning the code-generated API was not difficult for me so I don't really see the orm as a steep learning curve (although maybe a bit steeper when compared to the current proposal for collections layer).

I think the advantages of the orm are missing from this section and listing them might help a user understand when they would want to use the orm versus when they would want to use the collections layer. The advantages that have already been discussed in this pull request might be a good starting point.

We are using the orm in regen ledger and I don't see us switching as we have had no issues with the orm and we are leveraging the more complex functionality it provides to meet our needs, which includes multiple indexes, unique constraints on multiple indexes, and multi-column indexes.

Maybe some of this is planned for the collections layer but I would still prefer to define the state api using protobuf and to see protobuf more widely adopted for module development, "write once use forever".